File: packages/csharp-sdk/API/UsersClient.cs
using SchlepEngine.Types;

namespace SchlepEngine.API;

/// <summary>
/// Client for user operations.
/// </summary>
/// <remarks>
/// Provides methods for user profile management and API key operations.
/// </remarks>
/// <example>
/// <code>
/// var client = new SchlepClient("your-api-key");
/// var profile = await client.Users.GetProfileAsync();
/// Console.WriteLine($"User: {profile.Email}");
/// </code>
/// </example>
public class UsersClient
{
    private readonly SchlepClient _client;

    /// <summary>
    /// Initializes a new instance of the UsersClient class.
    /// </summary>
    /// <param name="client">The parent SchlepClient instance.</param>
    internal UsersClient(SchlepClient client)
    {
        _client = client;
    }

    /// <summary>
    /// Get current user profile.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>User profile information.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<UserProfile> GetProfileAsync(
        CancellationToken cancellationToken = default)
    {
        return await _client.GetAsync<UserProfile>("/users/me", cancellationToken);
    }

    /// <summary>
    /// Update user profile.
    /// </summary>
    /// <param name="updates">Fields to update.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Updated user profile.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<UserProfile> UpdateProfileAsync(
        object updates,
        CancellationToken cancellationToken = default)
    {
        return await _client.PatchAsync<UserProfile>("/users/me", updates, cancellationToken);
    }

    /// <summary>
    /// List API keys for the current user.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>List of API keys.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<List<ApiKeyInfo>> ListApiKeysAsync(
        CancellationToken cancellationToken = default)
    {
        var response = await _client.GetAsync<ApiKeysWrapper>("/users/me/api-keys", cancellationToken);
        return response.ApiKeys;
    }

    /// <summary>
    /// Create a new API key.
    /// </summary>
    /// <param name="name">API key name.</param>
    /// <param name="permissions">Optional permissions for the API key.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Created API key information.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<ApiKeyInfo> CreateApiKeyAsync(
        string name,
        string[]? permissions = null,
        CancellationToken cancellationToken = default)
    {
        var body = new { name, permissions = permissions ?? Array.Empty<string>() };
        return await _client.PostAsync<ApiKeyInfo>("/users/me/api-keys", body, cancellationToken);
    }

    /// <summary>
    /// Revoke an API key.
    /// </summary>
    /// <param name="keyId">API key ID.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>Revocation confirmation.</returns>
    /// <exception cref="ApiException">Thrown if the API returns an error.</exception>
    public async Task<object> RevokeApiKeyAsync(
        string keyId,
        CancellationToken cancellationToken = default)
    {
        return await _client.DeleteAsync<object>($"/users/me/api-keys/{keyId}", cancellationToken);
    }
}

File: labs/packages/java-sdk/src/main/java/io/schlepengine/api/UsersClient.java
package io.schlepengine.api;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.schlepengine.exceptions.ApiException;
import okhttp3.OkHttpClient;

import java.io.IOException;
import java.util.Map;

/**
 * Users API client.
 *
 * <p>Provides methods for user management and profile operations.</p>
 *
 * <h3>Features</h3>
 * <ul>
 *   <li>Get and update user profile</li>
 *   <li>Manage user preferences</li>
 *   <li>View user activity</li>
 *   <li>Manage API keys</li>
 * </ul>
 *
 * <h3>Example Usage</h3>
 * <pre>{@code
 * UsersClient usersClient = client.users();
 *
 * // Get current user profile
 * Map<String, Object> profile = usersClient.getProfile();
 *
 * // Update profile
 * Map<String, Object> updates = new HashMap<>();
 * updates.put("name", "New Name");
 * updates.put("email", "newemail@example.com");
 * Map<String, Object> updatedProfile = usersClient.updateProfile(updates);
 * }</pre>
 *
 * @author Schlep-engine Team
 * @version 1.0.0
 * @since 1.0.0
 */
public class UsersClient extends BaseClient {

    /**
     * Create a new users client.
     *
     * @param httpClient HTTP client instance
     * @param objectMapper JSON mapper instance
     * @param baseUrl API base URL
     * @param apiKey API authentication key
     */
    public UsersClient(
        OkHttpClient httpClient,
        ObjectMapper objectMapper,
        String baseUrl,
        String apiKey
    ) {
        super(httpClient, objectMapper, baseUrl, apiKey, "/users");
    }

    /**
     * Get current user profile.
     *
     * @return User information
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public Map<String, Object> getProfile() throws ApiException, IOException {
        JsonNode response = get("/me");
        return parseData(response, new TypeReference<Map<String, Object>>() {});
    }

    /**
     * Update user profile.
     *
     * @param updates Fields to update
     * @return Updated user information
     * @throws ApiException if the API returns an error
     * @throws IOException if there's a network or I/O error
     */
    public Map<String, Object> updateProfile(Map<String, Object> updates)
            throws ApiException, IOException {
        JsonNode response = patch("/me", updates);
        return parseData(response, new TypeReference<Map<String, Object>>() {});
    }
}